### PR TITLE
Low level integer class

### PR DESCRIPTION
**Pull request by isuruf**
WIth this PR you can use `piranha::integer` as the storage for `SymEngine::Integer`.
Use git master of Piranha for this.

Do `cmake -DWITH_PIRANHA=yes -DINTEGER_CLASS=piranha .`

- [x] Get all the functionality into piranha.
- [x] Remove dependence on libgmpxx
- [x] Enable polynomial and series tests

### CODE REVIEW

**Comment by certik** [comment]
Thanks for working on this. These changes look very good. Why is the `_z` literal needed? I.e. why did it work with gmp, but not with Piranha?

**Comment by isuruf** [comment]
Piranha has no implicit conversion from `int` to `piranha::integer`, so things like
```
std::vector<integer_class> a = {1, 2, 3};
```
were not working anymore.

Also I enabled C++11 in gmpxx and used std::move everywhere (around 15% improvement to speed of expand2)

**Comment by certik** [comment]
Ha, nice!

**Comment by isuruf** [comment]
This is working now. You need the Piranha git master to use `Piranha::integer` and `Piranha::rational` as the container for `Integer`.

**Comment by isuruf** [comment]
With `piranha::integer` as the integer class, running time for `expand2` is `863ms` while it's `1018ms` with `mpz_class` as the integer class

**Comment by isuruf** [comment]
@certik, can you review?

**Comment by certik** [comment]
Thanks for doing this. Very exciting. It's a huge patch, it will take me some time to go over this carefully. I went through it quickly so far, and things look good. I'll try to review it in detail soon.

**Comment by certik** [comment]
How does the Flint integer compare?

**Comment by isuruf** [comment]
I couldn't complete it due to some issues, see wbhart/flint2#222 and several other bugs.

**Comment by certik** [comment]
We need to write a wrapper on top of Flint's integer. That can be done
later.

Shouldn't the get_mpz_t and other functions have exactly the same return
type for both mpz_class and Piranha?

**Comment by isuruf** [comment]
`piranha::integer` returns a `mpz_view` class which is implicitly convertible to a `mpz_ptr`, but is tied to the lifetime of the `mpz_view` class. Therefore we cannot return a `mpz_ptr`

**Comment by certik** [comment]
I see, indeed we cannot.

Another question this: in this PR, the original mpz classes are used, and we create wrappers in terms of the `get_mpz_t` and similar functions. That's option 1). The option 2) is to require to write a wrapper class, that would have `get_mpz_t` as methods.

In case of Flint, if we want to get it working, we actually also have to write a wrapper class. In case of Piranha, my understanding is that you implemented the missing features. Ultimately though, we can't control other projects. So by default, we also have to write some kind of a wrapper. That would seem to speak to 2).

An argument for 1) is that `mpz_class` and flint (though we can't use it...) are using expression templates, thus making some syntax faster. However, as discussed in #478, you actually don't need the expression templates for speed. So if we simply make SymEngine work with an integer class as described in #478, then we don't need expression templates at all. However, the approach in #478 is still using non member functions to do the arithmetic, so that's essentially the approach 1).

Let me know what you think about the design.

File: symengine/mul.cpp [hunk 1]
```diff
@@ -456,10 +456,15 @@ RCP<const Basic> Mul::subs(const map_basic_basic &subs_dict) const
...
+        RCP<const Basic> factor_old;
+        if (eq(*p.second, *one)) {
+            factor_old = p.first;
+        } else {
+            factor_old = make_rcp<Pow>(p.first, p.second);
+        }
         RCP<const Basic> factor = factor_old->subs(subs_dict);
         if (factor == factor_old) {
-            Mul::dict_add_term_new(outArg(coef), d, p.second, p.first);
+            Mul::dict_add_term(d, p.second, p.first);
```
Comment (certik) [inline symengine/mul.cpp]: This seems unrelated to changing the integer class, or is it related? What was the motivation for this change? I am a little worried that changes like these might break things, and since this PR is so big, we don't want to be messing with such things, unless we have to.

If this is a good, but unrelated fix, then it should go into a separate PR, so that it's easy to revert if needed.

File: symengine/tests/printing/test_printing.cpp [hunk 2]
```diff
@@ -42,6 +42,12 @@ using SymEngine::complex_double;
 using SymEngine::BaseVisitor;
 using SymEngine::StrPrinter;
 using SymEngine::Sin;
+using SymEngine::integer_class;
+
+inline integer_class operator "" _z(const char* str)
+{
+    return integer_class(str);
+}
```
Comment (certik) [inline symengine/tests/printing/test_printing.cpp]: Later we probably want to move this into the SymEngine itself.

Comment (certik) [inline symengine/tests/printing/test_printing.cpp]: I can see that this is repeated in several tests, so this should be unified.

File: symengine/mul.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/mul.cpp]: Yes, you are right. This PR is big enough as it is. I'll remove this.

File: symengine/tests/printing/test_printing.cpp (hunk 2 above)
Comment (certik) [inline symengine/tests/printing/test_printing.cpp]: Actually, I can see that you did that already, but then moved it into the `.cpp` files in 79bde6db22a746ba5fdee88416c57d15bfff1d73, but the commit log does not explain why.

Comment (isuruf) [inline symengine/tests/printing/test_printing.cpp]: Yes, I moved it because it was a feature introduced in gcc-4.7 and at that time we were still using gcc-4.6 and I wanted check in Travis that it still compiles. I'll move this back to SymEngine code itself.

Comment (certik) [inline symengine/tests/printing/test_printing.cpp]: Perfect.

**Comment by isuruf** [comment]
@certik, about the design, I'd like to get a wrapper working with `mpz_t` or `fmpz_t` as the storage type. This would be as discussed in #478.
This PR shows that using `piranha::integer` do have benefits and I'm sure `fmpz_t` would give an even more speed benefit

**Comment by certik** [comment]
Ping me once you push in the patches fixing the two little things. I'll do a final review.

File: symengine/mp_class.h [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+ * mpz_t from libgmp and fmpz_wrapper which wraps fmpz_t from libflint. This choice
+ * is made at compile time with SYMENGINE_INTEGER_CLASS.
+ *
+ * Each of these classes has to have all the arithmetic operators overloaded with
+ * operands of these classes and C++ integer types except long long. Also, shift
+ * operators, move operators, string, integer and mpz_t constructors are also
+ * required.
+ *
+ * To add a new type, several non-member functions need to be defined for the new
+ * type. See mpz_wrapper implementation for all the non-member functions that needs
```
Comment (certik) [inline symengine/mp_class.h]: needs -> need

**Comment by certik** [comment]
I reviewed it again, I think it looks good, +1 to merge if tests pass.

**Comment by certik** [comment]
I merged another PR and now there are conflicts here. Sorry about that.

**Comment by isuruf** [comment]
Tests pass now

**Comment by certik** [comment]
I ran benchmarks, it looks like things didn't get any slower. The PR is +1 from me. I am merging it.

**Comment by certik** [comment]
Thanks for the work!
